presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizationManager.java

```diff
@@ -0,0 +1,268 @@
+        discoveryService.shutdown();
+    }
+
+    private void startDiscovery()
+    {
+        long startDelay = getStartDelay();
+
+        discoveryService.scheduleWithFixedDelay(() -> {
+            if (shutdown.get()) {
+                return;
+            }
+
+            try {
+                // jitter to avoid overloading database
+                long intervalSeconds = min(3600, (long) discoveryInterval.convertTo(SECONDS).getValue());
+                SECONDS.sleep(ThreadLocalRandom.current().nextLong(1, intervalSeconds));
+
+                log.info("Running shard organizer...");
+                discoverShardsAndSubmitJobs();
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+            catch (Throwable t) {
+                log.error(t, "Error running shard organizer");
+            }
+            finally {
+                shardOrganizerDao.updateShardOrganizerLastRunTime(currentNodeIdentifier);
+            }
+        }, startDelay, discoveryInterval.toMillis(), MILLISECONDS);
+    }
+
+    private long getStartDelay()
+    {
+        Long secondsAgo = shardOrganizerDao.getShardOrganizerLastRunSecondsAgo(currentNodeIdentifier);
+        if (secondsAgo == null) {
+            shardOrganizerDao.insertNode(currentNodeIdentifier);
+            secondsAgo = shardOrganizerDao.getShardOrganizerLastRunSecondsAgo(currentNodeIdentifier);
+        }
+        return max(discoveryInterval.toMillis() - SECONDS.toMillis(secondsAgo), 0);
+    }
+
+    @VisibleForTesting
+    void discoverShardsAndSubmitJobs()
+    {
+        Collection<Long> enabledTableIds = metadataDao.getOrganizationEnabledTables();
+        Set<ShardMetadata> allShards = shardManager.getNodeShards(currentNodeIdentifier);
+        ListMultimap<Long, ShardMetadata> tableShards = Multimaps.index(allShards, ShardMetadata::getTableId);
+
+        for (Map.Entry<Long, Collection<ShardMetadata>> entry : tableShards.asMap().entrySet()) {
+            long tableId = entry.getKey();
+
+            // If not enabled or not sorted, nothing to organize, skip the table
+            if (!enabledTableIds.contains(tableId) || metadataDao.listSortColumns(tableId).isEmpty()) {
+                continue;
+            }
+
+            Table tableInfo = metadataDao.getTableInformation(tableId);
+            Set<ShardMetadata> filteredShards = entry.getValue().stream()
+                    .filter(shard -> !organizer.inProgress(shard.getShardUuid()))
+                    .collect(toSet());
+
+            Collection<ShardIndexInfo> indexInfos = toShardIndexInfo(dbi, metadataDao, tableInfo, filteredShards, true);
+            Set<OrganizationSet> organizationSets = createOrganizationSets(tableInfo, indexInfos);
+
+            log.info("Created %s organization set(s) from %s shard(s) for table ID %s", organizationSets.size(), filteredShards.size(), tableId);
+            organizationSets.forEach(organizer::enqueue);
```

Comment:
Enqueue all of these after the discovery loop is complete. Otherwise, if there's a failure, you can't retry because you don't know what failed.

Or maybe we should keep track of organization per table, and possibly allow a different interval per table.